fcode : Improved printing of Fortran code

jashan498 opened the pull request:
Current implementation of `fcode` doesn't print different cases as different variable in Fortran code.
Example:
```
>>> fcode(X + 2*x*Y**Rational(7, 2))
'      X + 2*Y**(7.0d0/2.0d0)*x'
```
In this, though `X` and `x` were different variables in python but they both are same for Fortan.

So to change this behaviour, this patch appends `_` (till it's unique in the pool of words without considering case ) to similar symbols (but with a different case).
```
>>> fcode(X + 2*x*Y**Rational(7, 2))
'      X + 2*Y**(7.0d0/2.0d0)*x_'
```

jashan498 committed:
Improved printing of Fortan code

bjodah commented:
I like the idea (and the implementation). My only concern is that doing code-generation
probably involves calling `fcode` multiple times, and then I'm not sure this approach helps:
```
In [2]: from sympy import symbols, fcode

In [3]: X, x, x_ = symbols('X,x,x_')

In [4]: fcode(X*x*x_ + x*x_)
Out[4]: '      X*x_*x__ + x_*x__'

In [5]: fcode(X*x*x_)
Out[5]: '      X*x_*x__'

In [6]: fcode(x*x_)
Out[6]: '      x*x_'
```
That could lead to some hard to spot errors in larger bodies of code.

An alternative would be to instruct the user instantiate a FCodePrinter and then we add the "word_map" to the printer instance, that way separate invocations of `.doprint` will not generate
different names for the same symbols.

What do you think?

jashan498 commented:
@bjodah thanks for pointing out that!
I strongly agree with you that `fcode` should be consistent with what it prints every time. So, i made [these changes](https://github.com/sympy/sympy/compare/master...jashan498:oprint2l) in the code at another branch.
I made `word_map` global inside the `FCodePrinter ` class. So, now to use `fcode`, user have to  instantiate FCodePrinter with any random SymPy expression (or maybe a string of comma separated symbols) containing the symbols we intend to use. It would change `word_map` according to it and whenever we would use `fcode`, we will put that instance as an argument in it.
```
>>> from sympy.printing.fcode import FCodePrinter
>>> from sympy.printing.fcode import *
>>> X, x, x_ = symbols('X,x,x_')
>>> z = FCodePrinter(x + X + x_)
>>> fcode(z, X*x*x_ + x*x_)
'      X*x_*x__ + x_*x__'
>>> fcode(z, x*x_)
'      x_*x__'
>>> fcode(z, X*x*x_ )
'      X*x_*x__'
```

I wanted to make `fcode` a method of FCodePrinter class but i think it would not be consistent with other code_printers **maybe** . This is just a raw change, would commit a better version of this if this is the right way to do this : )

bjodah commented:
Looking at [that branch](https://github.com/sympy/sympy/compare/master...jashan498:oprint2#diff-fb4eda066a93d29487673f321eb3290d) I see that you would need to set the attribute `word_map` in `__init__` (otherwise modifications of it would be global). Also, you shouldn't update `word_map` in `__init__` but rather in `_print_Symbol`.

This has the downside that the exact renaming of e.g. `x` & `X` will depend on:
1. the traversal order of the tree being printed
2. order of `doprint` invocations.

..but I think we would have to live with it.

Renaming of Symbols in FCodePrinter is a good idea and I think it should become the default, but you should probably also introduce a printer setting e.g. "symbol_name_mangling" (default to ``True``) so that in case people already have "post-processors" written to handle this problem they can still use FCodePrinter (and just set the `symbol_name_mangling` flag to `False`).

smichr commented:
What about using the `disambiguate` function: X and x will become `x_1` and `x_2` (and I think it's smart enough to watch out for a case like `free` being `x, X`, and `x_1`).

```
free = list(expr.free_symbols)
lower_free = disambiguate([Symbol(i.name.lower()) for i in free])
lower_expr = expr.xreplace(dict(zip(free, lower_free)))
```

bjodah commented:
@smichr sounds like a good idea (if it be used to grow the substitution dictionary incrementally).

normalhuman commented:
@smichr My concern (perhaps unnecessary) was that the case of original names servedSomePurpose and the code might become less readable if the casing is lost, even though it's ignored by the compiler. 

Jashan didn't mention it, but testing also uncovered substitution issues around dummy variables (not identified by their names) and index variables (must be declared integer). This is why his implementation preserves original symbols when possible, and otherwise preserves their class and assumptions.

bjodah commented:
I think @normalhuman makes a fair point: x & X_ still tells you what their original names were (it does get complicated if for some reason the user had a symbol named `x_` or `X_` before). The printer option for mangling names could be allowed to be a string (where `True` would translate to `'_'`).

jashan498 commented:
Sorry for the late reply. 
Yeah, @normalhuman is right, it is indeed important to preserve the case.For cases involving, `x_`, `X_` and `x`, why not we add a function like `show_mapping()` for the user so that they can themselves see `word_map` if it gets complicated (in addition to changes @bjodah pointed out)?

bjodah commented:
No need for a separate function, just document the name of the attribute (I
suggest calling it "name_mangling").

On Jan 26, 2018 05:22, "Jashan" <notifications@github.com> wrote:

Sorry for the late reply.
Yeah, @normalhuman <https://github.com/normalhuman> is right, it is indeed
important to preserve the case.For cases involving, x_, X_ and x, why not
we add a function like show_mapping() for the user so that they can
themselves see word_map if it gets complicated (in addition to changes
@bjodah <https://github.com/bjodah> pointed out)?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/sympy/sympy/pull/14003#issuecomment-360680841>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAdqMDDJeyM6weFxm_pLKD3_becAeGWLks5tOVMhgaJpZM4RsBze>
.

jashan498 commented:
@bjodah, please see [this](https://github.com/sympy/sympy/compare/master...jashan498:oprint2) before I update all the `fcode` tests.
I added `word_map` inside `init`  and name mangling is done by `_print_Symbol` as you said (only when `name_mangling = True`)
```
>>> from sympy.printing.fcode import *
>>> from sympy.printing.fcode import FCodePrinter
>>> z = FCodePrinter()
>>> X_, x, x_ = symbols('X_,x,x_')
>>> fcode(z,x_*x*X_ + X_)
'      X_*x*x__ + X_'
>>> z.word_map
{X_: X_, x: x, x_: x__}
>>> fcode(z,x_*x*X_)
'      X_*x*x__'
>>> z.word_map
{X_: X_, x: x, x_: x__}
>>> fcode(z,x_*x*X_, false)
'      X_*x*x_'
```

bjodah commented:
@jashan498 that's a good start, here are a few remarks of mine on the code in your aforementioned link:
1. Override `_print_Symbol` in `FCodePrinter` instead.
2. Rename `word_map` to something more descriptive e.g. `mangled_symbols`
3. You shouldn't change `fcode`'s body (look at how other printer options are implemented)

jashan498 commented:
@bjodah, I will implement all these but Isn't it better that name mangling for every language is done by `_print_Symbol` of CodePrinter only as it is parent class for all than overriding it?

bjodah commented:
@jashan498 I can't think of any other languages we support which are case-sensitive (I might be wrong though).

jashan498 commented:
Maybe you are right, its more appropriate to override `_print_Symbol` separately only if needed. 
Btw where according to you is ideal place for `fob.used_name= []` (important to initialize to an empty list every time fcode is called ) and ` fob.symbol_name_mangling = name_mangling` (can be placed inside `init` though), if we cant place it inside `fcode` ?

bjodah commented:
You can initialize the mapping in `__init__`. An ordinary dictionary will do (I think). The printer setting is introduced by adding `'name_mangling': True` to `FCodePrinter._default_settings`. Introspection of this variable is then done against `self._settings['name_mangling']` in `FCodePrinter._print_Symbol`.

jashan498 commented:
I changed `FCodePrinter` according to what @bjodah said and updated syntax of all `fcode` tests inside `test_fcode.py` but I'm getting an error on `test/doctest printing` command insidev`sympy/doc/src/modules/printing.rst` because apparently i am supposed to change syntax of `fcode` calling there too. 
I have never updated anything inside `src` directory. Is it fine to do so ?

jashan498 committed:
transfered all the new code to _print_Symbol

jashan498 commented:
@normalhuman @bjodah @smichr please review are these changes appropriate?

bjodah commented:
@jashan498 you should not change the signature of `fcode`. (it will create a printer so no need pass an instance)

jashan498 commented:
@bjodah the only reason I am passing an instance is for previously used  `mangled_symbols` (word_map) :  
`return FCodePrinter(settings, fob.mangled_symbols).doprint(expr, assign_to)`
inside `fcode`, so that the mapping remain same everytime someone use fcode with that instance atleast.

If the signature of `fcode` shouldn't be changed, can I make that instance an optional argument:
`def fcode(expr, assign_to = None, **settings,  fob = FCodePrinter()) `
This way we don't need to change the signature of fcode ?

bjodah commented:
If we fix it in such a way that:
```
FCodePrinter(settings).doprint(expr)
```
does the right thing, then `fcode` will also work without us touching it.

jashan498 commented:
How about making `mangled_symbols` global to `FCodePrinter` ? 
Then there would be no need to create instance too, everything would be done inside `fcode` only and everytime same result would be produced for same symbols.

bjodah commented:
Relying on mutable global state (be it bound to a class or not) should really only be the last resort (you can imagine a user using say 2 libraries both using `fcode`, those may then interact unexpectedly). And here we don't need to. Setting up the mutable mapping on the instance in `__init__` and modifying it in `_print_Symbol` solves the problem and avoids global state.

jashan498 committed:
removed necessity of instantiating of FCodePrinter

jashan498 commented:
@bjodah, I did the required changes and its working fine now with the same signature. Tell if there is more scope for further improvements. Here is one example for this commit :
```
>>> X, x = symbols('X ,x')
>>> fcode(x - X ,assign_to='me')
'      me = -X + x_'
>>> fcode(x ,assign_to='me')
'      me = x_'
>>> fcode(X ,assign_to='me')
'      me = X'
```

bjodah commented on sympy/printing/fcode.py:
```diff
@@ -103,7 +104,9 @@ class FCodePrinter(CodePrinter):
 
-    def __init__(self, settings={}):
+    def __init__(self, settings={}, mangled_symbols={}):
+        self.mangled_symbols = mangled_symbols         ## Dict showing mapping of all words
```
This needs to read `self.mangled_symbols = {}`. (drop `mangled_symbols` as keyword argument)

bjodah commented on sympy/printing/fcode.py:
```diff
@@ -129,6 +131,33 @@ def _lead(self):
             raise ValueError("Unknown source format: %s" % self._settings['source_format'])
 
+    def _print_Symbol(self, expr):
+        if self._settings['name_mangling'] is True:
```
`is True` is problematic (e.g. `(numpy.array([1]) == 1)[0] is True` returns `False` for example)

bjodah commented on sympy/printing/fcode.py:
```diff
@@ -129,6 +131,33 @@ def _lead(self):
 
+    def _print_Symbol(self, expr):
+        if self._settings['name_mangling'] is True:
+            from sympy import symbols, ordered, Symbol
```
do these imports need to be here? (or would e.g. `from sympy.core.symbol import Symbol` work at the beginning of this file?)

bjodah commented on sympy/printing/fcode.py:
```diff
@@ -129,6 +131,33 @@ def _lead(self):
+                    if name == sym.name:
+                        self.mangled_symbols[sym] = sym
+                    else:
+                        self.mangled_symbols[sym] = symbols(name, cls=sym.__class__, **sym.assumptions0)
```
just wondering: do we need `symbols` here? (or can we simply use `Symbol`?)

jashan498 replied:
I tried `Symbol` here but it is not accepting these arguments.
It gives `TypeError: __new__() got multiple values for argument 'cls'`

normalhuman replied:
The reason being, if sym is of class Idx (for example), then the replacement should also be Idx. The function `symbols` is able to create such things since it takes cls argument.

bjodah replied:
Oh yes, forgot about that. We should probably add a test for that specific case so that it won't be overlooked in e.g. a future refactoring.

jashan498 replied:
@bjodah  [this](https://github.com/jashan498/sympy/blob/1562b0e824129f6f00a500ab954f2febe7d2f83d/sympy/printing/tests/test_fcode.py#L187) assures that. It gives assertion error without it but still i am adding one more for that.

bjodah commented on sympy/printing/fcode.py:
```diff
@@ -506,7 +535,7 @@ def indent_code(self, code):
 
 
-def fcode(expr, assign_to=None, **settings):
+def fcode(expr, assign_to = None, **settings):
```
no need to put spaces around `=`

bjodah commented on sympy/printing/fcode.py:
```diff
@@ -570,6 +604,16 @@ def fcode(expr, assign_to=None, **settings):
+    >>> Z, z = symbols('Z,z')
+    >>> fcode(Z - z)
+    '      Z - z_'
+    >>> fcode(Z - z, name_mangling= False)
```
nit picking (not important): `name_mangling=False`

jashan498 replied:
my output is inconsistent for `python2` and `python3`, you have any clue why?
python3 : 
```
Python 3.5.2 (default, Nov 23 2017, 16:37:01) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.printing.tests.test_fcode import *
>>> Z, z = symbols('Z,z')
>>> fcode(Z - z)
'      Z - z_'
```
python2:
```
Python 2.7.12 (default, Nov 20 2017, 18:23:56) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.printing.tests.test_fcode import *
>>> Z, z = symbols('Z,z')
>>> fcode(Z - z)
'      Z - z'
```

bjodah commented on sympy/printing/tests/test_fcode.py:
```diff
@@ -75,7 +74,15 @@ def test_fcode_Float():
+def test_case():
+    x,x_,x__,y,X,X_,Y = symbols('x,x_,x__,y,X,X_,Y')
+    assert fcode(exp(x_) + sin(x*y) + cos(X*Y)) == \
+                        '      exp(x_) + sin(x*y) + cos(X*Y)'
```
three versions of "x" should give x with 0, 1 & 2 underscores in output right?

bjodah commented on sympy/printing/tests/test_fcode.py:
```diff
@@ -75,7 +74,15 @@ def test_fcode_Float():
+                        '      exp(x_) + sin(x*y) + cos(X*Y)'
+
+    assert fcode(exp(x__) + 2*x*Y*X_**Rational(7, 2)) == \
+                        '      2*X_**(7.0d0/2.0d0)*Y*x + exp(x__)'
```
We should add tests for fcode(..., name_mangling=False) also. And some tests for an instance of FCodePrinter.

bjodah reviewed (changes requested):
That keyword argument in `__init__` is actually mutable global state (one of Python's "misfeatures").

bjodah commented:
@jashan498 regarding your example in your previous comment, I'd expect it to work like this:
```
>>> X, x = symbols('X, x')
>>> fcode(x - X, assign_to='me')
'      me = -X + x_'
>>> fcode(x ,assign_to='me')
'      me = x'
>>> fcode(X ,assign_to='me')
'      me = X'
```
For it to behave the way you wrote the user would need to instantiate an instance of the printer:
```
>>> X, x = symbols('X ,x')
>>> p = FCodePrinter()
>>> p.doprint(x - X ,assign_to='me')
'      me = -X + x_'
>>> p.doprint(x ,assign_to='me')
'      me = x_'
>>> p.doprint(X ,assign_to='me')
'      me = X'
```

normalhuman commented on sympy/printing/fcode.py:
```diff
@@ -546,6 +574,11 @@ def fcode(expr, assign_to=None, **settings):
         more features are added.
+    name_mangling : bool, optional
+        If True, then variables (which would have been identical in
+        case-insensitive Fortan) are mangled by appending different number
```
It's Fortran, not Fortan. (I corrected that elsewhere). Parentheses don't quite work here. I suggest "If True, then the variables that would become identical in case-insensitive Fortran are..."

jashan498 committed:
Few changes inside FCodePrinter. Also added few tests for and invoking by an instance

jashan498 commented:
@bjodah I have no idea why one build test is failing (it passes all locally).

jksuom commented on sympy/printing/fcode.py:
```diff
@@ -546,6 +574,11 @@ def fcode(expr, assign_to=None, **settings):
+    name_mangling : bool, optional
+        If True, then the variables that would become identical in
+        case-insensitive Fortran are mangled by appending different number
+        of `_` at the end. If False, SymPy won't interfere with naming of
```
The error message
```
l.85856 of \(_\)
                 at the end. If False, SymPy wonâ€™t interfere with naming of
```
refers to this line. Sphinx is configured to replace the pair of backquotes by `\(` and `\)`, and Latex cannot interpret the resulting `\(_\)`. Double backquotes could probably be used.

jashan498 replied:
thanks!

jashan498 committed:
added test for indexed objects

bjodah commented on sympy/printing/tests/test_fcode.py:
```diff
@@ -89,6 +89,14 @@ def test_case():
     assert ob.doprint(X*sin(x), assign_to='mu') == '      mu = X*sin(x_)'
     assert ob.doprint(x_, assign_to='ad') == '      ad = x__'
+    A = IndexedBase('A')
+    i = Idx('i', symbols('n', integer=True))
```
could we have a test with two indices, e.g. `i` and `I` to make (really) sure name mangling works for instances of `Idx`?

bjodah commented on sympy/printing/tests/test_fcode.py:
```diff
@@ -75,7 +74,29 @@ def test_fcode_Float():
+    assert ob.doprint(x_, assign_to='ad') == '      ad = x__'
+    A = IndexedBase('A')
+    i = Idx('i', symbols('n', integer=True))
+    sq = implemented_function('sq', Lambda(x, x**2))
```
Is this needed? Or can we just have `fcode(A[i]**2, assign_to=A[i])` below? (or rather something with two indices `I` and `i`)

bjodah commented:
Apart from my two comments, this looks ready to me.

jashan498 committed:
updated test_case for Idx instance

ylemkimon commented on sympy/printing/fcode.py:
```diff
@@ -129,6 +131,32 @@ def _lead(self):
 
+    def _print_Symbol(self, expr):
+        if self._settings['name_mangling'] == True:
+            for sym in tuple(ordered(expr.atoms(Symbol))):
```
`expr` is an instance of `Symbol`

bjodah replied:
Yes, he is absolutely right (I missed this -- that's why multiple reviewers is a good thing). Yes, go ahead and simplify this logic as much as you can.

ylemkimon commented on sympy/printing/fcode.py:
```diff
@@ -23,8 +23,8 @@
-from sympy.core import S, Add, N, Float
-from sympy.core.compatibility import string_types, range
+from sympy.core import S, Add, N, Float, Symbol, symbols
+from sympy.core.compatibility import string_types, range, ordered
```
`ordered` is no longer used

jashan498 replied:
Deleted. even `Symbol` was unnecessary.

ylemkimon commented on sympy/printing/fcode.py:
```diff
@@ -129,6 +131,31 @@ def _lead(self):
+                if name == expr.name:
+                    self.mangled_symbols[expr] = expr
+                else:
+                    self.mangled_symbols[expr] = symbols(name, cls=expr.__class__, **expr.assumptions0)
```
Why not use `Symbol`?

jashan498 replied:
`Symbol` was unable to make `name` an `Idx` instance.

jashan498 replied:
I think after shifting the code from `fcode` to inside of `_print_Symbol`,  the need for `cls=expr.__class__`  and `**expr.assumptions0` has vanished (i don't know why) and `Symbol` is working too now!

jashan498 committed:
removed one loop and replaced symbols with Symbol

ylemkimon commented on sympy/printing/fcode.py:
```diff
@@ -129,6 +131,31 @@ def _lead(self):
+
+        name = super(CodePrinter, self)._print_Symbol(expr)
+
+        if name in self.reserved_words:
```
Is this part of this PR?

jashan498 replied:
yeah, as this method is overriding `_print_Symbol` of `CodePrinter` class, it should also perform checks done by that method.

jashan498 replied:
or rather there is no need of it because `name = super(CodePrinter, self)._print_Symbol(expr)` is calling it anyway ?

jashan498 replied:
@ylemkimonI just checked, overriding `print_Symbol` is indeed avoiding [this](https://github.com/sympy/sympy/blob/888f16155784747383083a8cd48a7f0089780899/sympy/printing/codeprinter.py#L325). So that check has to be added inside `FcodePrinter.print_Symbol` too.

ylemkimon replied:
You have to pass the subclass as the first parameter of `super`, like `name = super(FCodePrinter, self)._print_Symbol(expr)`.

jashan498 committed:
changed arguments of super

normalhuman commented:
This looks like it was thoroughly discussed, any objections of comments before merging?

ylemkimon commented on sympy/printing/fcode.py:
```diff
@@ -129,6 +131,31 @@ def _lead(self):
+        if name in self.reserved_words:
+            if self._settings['error_on_reserved']:
+                msg = ('This expression includes the symbol "{}" which is a '
+                       'reserved keyword in this language.')
```
This is no longer needed, since it's called in the super method.

jashan498 committed:
removed redundant check on reserved words

normalhuman commented:
Good catch for the now-redundant check.
